### ISSUE TITLE
Add more examples on generic restrictions

### ISSUE THREAD

**Issue by danielo515** [facebook/flow#8569]
Generics are way more powerful than I expected. You can not only use them to take any argument, but you can also add restrictions on them. However, this is not clearly documented: the explanation is too short and the provided examples are too simplistic to give you an idea of what is possible.
I'm talking about[ this specific part](https://flow.org/en/docs/types/generics/#generic-types-act-as-bounds-).
Reading the doc and the provided example `<T: string>` I can't understand what's the point on that. To me it looks that you can take a generic that must be a string, which feels redundant: why not just type it as string?

I think that a much better example (or at least, a good extra example) would be something like `<T: {id: string}>` and then explain that you can take any type as argument as long as it has an `id` property.

As an example of this confusion, take a look at this stack-overflow question:

https://stackoverflow.com/questions/65737073/flow-type-a-generic-function-that-groups-any-type-that-has-at-least-an-id/65738542#65738542

Regards

**Comment by gkz** [comment]
You can have a type that is more specific than `string`, take a look at https://flow.org/en/docs/types/literals/

For example, the type `"foo"` is a type with only a single valid value, the string `"foo"`, and it is a subtype of `string`. 

More useful is when you have unions of these literal types, for example `"foo" | "bar"` (e.g. could represent the keys of an object). So a union of string literal types is a subtype of `string`.

I agree the documentation could be better.

**Comment by danielo515** [comment]
Yes, I said that feels redundant, my wording could have been better. What I try to express is that it is not easy to understand unless you are familiar with the use cases you mentioned. I think the example of objects containing "at least" this prop, but still allowing any type is easier to grasp for the majority of the developers.
